lookout/core/metrics/server.py

```diff
@@ -0,0 +1,225 @@
+    """It stores the count, the cumulative sum and, the cumulative sum of squares of the
+    observed variable. It makes it possible to calculate the rolling mean and standard deviation
+    of a time series.
+    """
+
+    def _metric_init(self):
+        super(RollingStats, self)._metric_init()
+        self._square = values.ValueClass(
+            self._type,
+            self._name,
+            self._name + "_sum_of_squares",
+            self._labelnames,
+            self._labelvalues,
+        )
+
+    def observe(self, amount):
+        """Observe the given amount."""
+        super(RollingStats, self).observe(amount=amount)
+        self._square.inc(amount ** 2)
+
+    def _child_samples(self):
+        count = max(self._count.get(), 1)
+        _sum = self._sum.get()
+        square = self._square.get()
+        return ("_count", {}, count), ("_sum", {}, _sum), ("_sum_of_squares", {}, square)
+
+
+def start_prometheus(port: int, addr: str = "", **kwargs):
+    """
+    Start the prometheus HTTP Server in the target port and address. The stored metrics will be
+    accessible at http://addr:port.
+
+    :param port: Target port to run the PrometheusServer.
+    :param addr: IP address of the PrometheusServer. It defaults to locahost.
+    :param kwargs: Additional keyword arguments to initialize the server. Not used by default.
+    :return: None
+    """
+    global PROMETHEUS_SERVER
+    PROMETHEUS_SERVER = PrometheusServer(port=port, addr=addr, **kwargs)
+    PROMETHEUS_SERVER.start_http_server()
+
+
+def submit_event(
+    key: str,
+    value: Union[int, float, bool],
+    labelnames: str = "",
+    metric_type: Callable = RollingStats,
+    update_method=None,
+    *args,
+    **kwargs,
+):
+    """Register an event by a key and with a numeric value. If the key does not exist, it creates a
+    new Prometheus Metric.
+
+    :param key: Identifier of the event.
+    :param value: Value of the event. It will convert cast variables to int.
+    :param labelnames: Additional description of the event. Only used when creating a new event.
+    :param metric_type: Type of metric that will be used to register the target value. Only used
+        when creating a new event.
+    :param update_method: Function to be called on the target metric to update the values. If None
+        it will try to apply update the metric calling observe(), inc() if observe is not
+        available, and set() if inc is not available.
+    :return: None
+    """
+    if PROMETHEUS_SERVER is not None:
+        PROMETHEUS_SERVER.submit_event(
+            key=key,
+            value=value,
+            labelnames=labelnames,
+            metric_type=metric_type,
+            update_method=update_method,
+            *args,
+            **kwargs,
+        )
+    else:
+        raise ValueError(
+            "PrometheusServer is not started yet, please call start_prometheus() first."
+        )
+
+
+class PrometheusServer:
```

Comment:
The metric registry should be outside of the Prometheus server. The Prometheus server should be the default class from the official library, no changes. The registry just holds the existing metrics, which we retrieve inside `submit_event`.

Reply:
If we remove this, we wont be able to make any future changes to the class, and the metrics will always be registered on the default prometheus registry.

This also makes it impossible to use other endpoints rather than the simple HttpServer (No push gateway, or no other prometheus_client functionality, because for all other use cases you need to manually specify the target registry.

Reply:
Totally fine